<?xml version="1.0"?>

<launch>

    <!-- extern parameters -->
    <arg name="output" default="screen" />
    <arg name="tfFixedFrame" default="map" />
    <arg name="tfBaseFootprintFrame" default="base_footprint" />
    <arg name="tfOdomFrame" default="odom" />
    <arg name="tfOutputFrame" default="ohm_pf_output" />
    <arg name="topOdometry" default="robot0/odom" />
    <arg name="top2dPoseEst" default="initialpose" />
    <arg name="topCeilCam" default="ceilCamPoseArray" />
    <arg name="topParticleCloud" default="particlecloud" />
    <arg name="topProbPose" default="probPose" />
    <arg name="topMap" default="map" />
    <arg name="topMapSrv" default="static_map" />
    <arg name="samplesMax" default="5000" />
    <arg name="samplesMin" default="50" />
    <arg name="topScan" default="robot0/laser_0" />
    <arg name="maxDistanceProbMap" default="10" />
    <arg name="subsamplingRateLaser" default="3" />
    <arg name="resamplingIntervallFilter" default="0.5" />
    <arg name="outputIntervallFilter" default="0.2" />
    <arg name="uncertaintyLaser" default="0.5" />
    <arg name="resamplingMethod" default="STD" />
    <arg name="initX" default="0.0" />
    <arg name="initY" default="0.0" />
    <arg name="initPhi" default="0.0" />
    <arg name="initSigmaTrans" default="0.5" />
    <arg name="initSigmaRot" default="0.2" />
    <arg name="initMode" default="GL" />
    <arg name="OCSThresholdLaser" default="0.2" />
    <arg name="OCSThresholdOdom" default="0.001" />
    <arg name="OCSThresholdResampler" default="0.2" />
    <arg name="skipParticleForGui" default="10" />
    <arg name="countLasers" default="1" />
    <arg name="filterLoopRate" default="20" />
    
    <!-- ******************* -->
    <!-- For Debugging: <node name="ohm_pf_node" pkg="ohm_pf" type="OhmPfNode" output="$(arg output)" launch-prefix="xterm -e cgdb -ex run -->  <!--args": append this; xml comment problem here -->
    <node name="ohm_pf_node" pkg="ohm_pf" type="OhmPfNode" output="$(arg output)" >
        <param name="tfFixedFrame" type="string" value="$(arg tfFixedFrame)" />
        <param name="tfBaseFootprintFrame" type="string" value="$(arg tfBaseFootprintFrame)" />
        <param name="tfOdomFrame" type="string" value="$(arg tfOdomFrame)" />
        <param name="tfOutputFrame" type="string" value="$(arg tfOutputFrame)" />
        <param name="topOdometry" type="string" value="$(arg topOdometry)" />
        <param name="top2dPoseEst" type="string" value="$(arg top2dPoseEst)" />
        <param name="topCeilCam" type="string" value="$(arg topCeilCam)" />
        <param name="topParticleCloud" type="string" value="$(arg topParticleCloud)" />
        <param name="topProbPose" type="string" value="$(arg topProbPose)" />
        <param name="topMap" type="string" value="$(arg topMap)" />
        <param name="topMapSrv" type="string" value="$(arg topMapSrv)" />
        <param name="uncertaintyLaser" type="double" value="$(arg uncertaintyLaser)" />
        <param name="resamplingMethod" type="string" value="$(arg resamplingMethod)" />
        <param name="samplesMax" type="int" value="$(arg samplesMax)" />
        <param name="samplesMin" type="int" value="$(arg samplesMin)" />
        <param name="topScan" type="string" value="$(arg topScan)" />
        <param name="maxDistanceProbMap" type="int" value="$(arg maxDistanceProbMap)" />
        <param name="subsamplingRateLaser" type="int" value="$(arg subsamplingRateLaser)" />
        <param name="countLasers" type="int" value="$(arg countLasers)" />
        <param name="resamplingIntervallFilter" type="double" value="$(arg resamplingIntervallFilter)" />
        <param name="outputIntervallFilter" type="double" value="$(arg outputIntervallFilter)" />
        <param name="filterLoopRate" type="double" value="$(arg filterLoopRate)" />
        <param name="skipParticleForGui" type="int" value="$(arg skipParticleForGui)" />

        <param name="initX" type="double" value="$(arg initX)" />
        <param name="initY" type="double" value="$(arg initY)" />
        <param name="initPhi" type="double" value="$(arg initPhi)" />
        <param name="initSigmaTrans" type="double" value="$(arg initSigmaTrans)" />
        <param name="initSigmaRot" type="double" value="$(arg initSigmaRot)" />
        <param name="initMode" type="string" value="$(arg initMode)" />
        
        <param name="OCSThresholdLaser" type="double" value="$(arg OCSThresholdLaser)" />
        <param name="OCSThresholdOdom" type="double" value="$(arg OCSThresholdOdom)" />
        <param name="OCSThresholdResampler" type="double" value="$(arg OCSThresholdResampler)" />
        
        <param name="OCSRotToTransFactor" type="double" value="8.0" />
        <param name="minimumValidScanRaysFactor" type="double" value="0.5" />
        <param name="additionalTranslationalNoise" type="double" value="0.05" />
        <param name="additionalRotationalNoise" type="double" value="0.17" /> <!-- 10.0 / 180.0 * M_PI -->
        <param name="lowVarianceFactor" type="int" value="3" />
    </node>

</launch>
